BitFaster.Caching/ScopedCache.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BitFaster.Caching
+{
+    // what happens if we completely encapsulate scoped?
+    // we can't implement ICache, since return types will now be Lifetime, not T
+    public class ScopedCache<K, T> : IScopedCache<K, T> where T : IDisposable
+    {
+        private readonly ICache<K, Scoped<T>> innerCache;
+
+        public ScopedCache(ICache<K, Scoped<T>> innerCache)
+        {
+            this.innerCache = innerCache;
+        }
+
+        public bool TryGet(K key, out Lifetime<T> value)
+        {
+            if (innerCache.TryGet(key, out var scope))
+            {
+                if (scope.TryCreateLifetime(out var lifetime))
+                {
+                    value = lifetime;
+                    return true;
+                }
+            }
+
+            value = default(Lifetime<T>);
+            return false;
+        }
+
+        // If we completely encapsulate Scoped, then we must allocate a new value factory for every GetOrAdd call
+        public Lifetime<T> GetOrAdd(K key, Func<K, Scoped<T>> valueFactory)
```

Comment:
This isn't totally fool proof - if the caller's delegate returns a Scoped instance that is disposed, the GetOrAdd methods will infinite loop.

There is no obvious way to check this, since the delegate is called by ICache, so ScopedCache can't see it. By the time ScopedCache sees the value, it has no way to differentiate between a race where Scoped got disposed, and a deterministically broken value factory.

Reply:
Options for safe scopes are:

1. Let the user screw it up, but have best perf
2. For debug only, insert an additional layer and verify new scoped instances are not disposed
3. Always insert the additional layer, suck up the extra delegate allocs
4. Force user to implement a factory class, factory enforces the constraint